api/utils/sshutils/conn_test.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sshutils
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"net"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/gravitational/teleport/api/constants"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/crypto/ssh"
+)
+
+type server struct {
+	listener net.Listener
+	config   *ssh.ServerConfig
+	handler  func(*ssh.ServerConn)
+	t        *testing.T
+	mu       sync.RWMutex
+	closed   bool
+
+	cSigner ssh.Signer
+	hSigner ssh.Signer
+}
+
+func (s *server) Run() {
+	for {
+		conn, err := s.listener.Accept()
+
+		s.mu.RLock()
+		if s.closed {
+			s.mu.RUnlock()
+			return
+		}
+		s.mu.RUnlock()
+
+		require.NoError(s.t, err)
+
+		go func() {
+			defer conn.Close()
+			sconn, _, _, err := ssh.NewServerConn(conn, s.config)
+			require.NoError(s.t, err)
```

Comment:
Using `require` within a goroutine is a bad idea. If the assertion fails it calls `t.FailNow` and according to the docs shouldn't be called from spawned goroutines:

> FailNow marks the function as having failed and stops its execution by calling runtime.Goexit (which then runs all deferred calls in the current goroutine). Execution will continue at the next test or benchmark. FailNow must be called from the goroutine running the test or benchmark function, not from other goroutines created during the test. Calling FailNow does not stop those other goroutines. 
https://pkg.go.dev/testing#T.FailNow

Reply:
Whoops didn't realize this. Updated [a59074c](https://github.com/gravitational/teleport/pull/11875/commits/a59074c1320f721943e2acd3ddc04117825547de)